### ISSUE TITLE
Pulumi terminates EC2 instance if security groups are added or removed

### ISSUE THREAD

**Issue by hseagle2015** [pulumi/pulumi#4741]
I'm using Pulumi 2.3.0 and Pulumi Python SDK to deploy an EC2 instance with one security group attached. If I attach additional security group to that same instance, Pulumi terminates previously create EC2 instance and deploys a new instance with 2 security groups.

Is it possible to avoid EC2 instance termination and instruct Pulumi to simply attach additional security group to existing instance?

If I enable EC2 instance termination protection (by passing `disable_api_termination=True` parameter to aws.ec2.Instance() function), then I get the following error from Pulumi:

```
(venv) user@host:~/pulumi-workspace$ pulumi up
Previewing update (myproject-p1-dev):
     Type                 Name                     Plan       Info
     pulumi:pulumi:Stack  myproject-p1-myproject-p1-dev
 ~   └─ aws:ec2:Instance  myproject-ec2-test          update     [diff: ~securityGroups]

Outputs:
  ~ Public IP      : "XXX.XXX.XXX.XXX" => output<string>
  ~ Public hostname: "ec2-XXX-XXX-XXX-XXX.eu-north-1.compute.amazonaws.com" => output<string>

Resources:
    ~ 1 to update
    3 unchanged

Do you want to perform this update? yes
Updating (myproject-p1-dev):
     Type                 Name                     Status                  Info
     pulumi:pulumi:Stack  myproject-p1-myproject-p1-dev  **failed**              1 error; 22 messages
 ~   └─ aws:ec2:Instance  myproject-ec2-test          **updating failed**     [diff: ~securityGroups]; 1 error

Diagnostics:
  pulumi:pulumi:Stack (myproject-p1-myproject-p1-dev):
    panic: fatal: An assertion has failed: Expected diff to not require deletion or replacement during Update of urn:pulumi:myproject-p1-dev::myproject-p1::aws:ec2/instance:Instance::myproject-ec2-test
    goroutine 28 [running]:
    github.com/pulumi/pulumi/sdk/v2/go/common/util/contract.failfast(...)
        /home/travis/gopath/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.1.1-0.20200508232528-aa313aecf8a0/go/common/util/contract/failfast.go:23
    github.com/pulumi/pulumi/sdk/v2/go/common/util/contract.Assertf(0xc000be2000, 0x5ddd59b, 0x48, 0xc000ee1728, 0x1, 0x1)
        /home/travis/gopath/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.1.1-0.20200508232528-aa313aecf8a0/go/common/util/contract/assert.go:33 +0x197
    github.com/pulumi/pulumi-terraform-bridge/v2/pkg/tfbridge.(*Provider).Update(0xc0003d3680, 0x673b600, 0xc000e9af00, 0xc00022a000, 0xc0003d3680, 0x5127c01, 0xc000e22280)
        /home/travis/gopath/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v2@v2.3.1/pkg/tfbridge/provider.go:825 +0x81d
    github.com/pulumi/pulumi/sdk/v2/proto/go._ResourceProvider_Update_Handler.func1(0x673b600, 0xc000e9af00, 0x5b31960, 0xc00022a000, 0x5b28360, 0xa2ce808, 0x673b600, 0xc000e9af00)
        /home/travis/gopath/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.1.1-0.20200508232528-aa313aecf8a0/proto/go/provider.pb.go:1920 +0x89
    github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x673b600, 0xc000e9a000, 0x5b31960, 0xc00022a000, 0xc000b440e0, 0xc000b44100, 0x0, 0x0, 0x6683140, 0xc000409170)
        /home/travis/gopath/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20180507213350-8e809c8a8645/go/otgrpc/server.go:57 +0x2f9
    github.com/pulumi/pulumi/sdk/v2/proto/go._ResourceProvider_Update_Handler(0x5bdd220, 0xc0003d3680, 0x673b600, 0xc000e9a000, 0xc0005b8060, 0xc00048be80, 0x673b600, 0xc000e9a000, 0xc000e90000, 0x75c)
        /home/travis/gopath/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.1.1-0.20200508232528-aa313aecf8a0/proto/go/provider.pb.go:1922 +0x14b
    google.golang.org/grpc.(*Server).processUnaryRPC(0xc0007241a0, 0x67606c0, 0xc000572a80, 0xc00122e200, 0xc000915ce0, 0xa2a0898, 0x0, 0x0, 0x0)
        /home/travis/gopath/pkg/mod/google.golang.org/grpc@v1.28.0/server.go:1082 +0x4fd
    google.golang.org/grpc.(*Server).handleStream(0xc0007241a0, 0x67606c0, 0xc000572a80, 0xc00122e200, 0x0)
        /home/travis/gopath/pkg/mod/google.golang.org/grpc@v1.28.0/server.go:1405 +0xd23
    google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0009b6050, 0xc0007241a0, 0x67606c0, 0xc000572a80, 0xc00122e200)
        /home/travis/gopath/pkg/mod/google.golang.org/grpc@v1.28.0/server.go:746 +0xbb
    created by google.golang.org/grpc.(*Server).serveStreams.func1
        /home/travis/gopath/pkg/mod/google.golang.org/grpc@v1.28.0/server.go:744 +0xa1

    error: update failed

  aws:ec2:Instance (myproject-ec2-test):
    error: transport is closing

Outputs:
  - Public IP      : "XXX.XXX.XXX.XXX"
  - Public hostname: "ec2-XXX-XXX-XXX-XXX.eu-north-1.compute.amazonaws.com"
  - ec2-test       : "sg-XXXXXXXXXXXXXXXXX"
  - pulumi-my-sg1  : "sg-XXXXXXXXXXXXXXXXX"
  - myproject-ec2-test: "i-XXXXXXXXXXXXXXXXX"

Resources:
    3 unchanged

Duration: 6s

```

**Comment by stack72** [comment]
Hi @hseagle2015 

Thanks for opening an issue here. I am not able to recreate this locally - please can you try and recreate it and show me the detailed diff that Pulumi gives you in the "details" view of the CLI?

This would be able to help me understand what is going on for you

Thanks

Paul

**Comment by clstokes** [comment]
@hseagle2015 I think this is a “classic ec2” nuance. See https://github.com/pulumi/pulumi-aws/issues/852 for a discussion about it. 

The fix is to use `vpcSecurityGroupIds` rather than `securityGroups` on the instance. 

@stack72 , it looks like the deprecation message is not working?

**Comment by stack72** [comment]
@clstokes there are no depreciations in Python :/ the docs should show it though

**Comment by stack72** [comment]
Let me rephrase that - there are no parameter level depreciations in Python

**Comment by hseagle2015** [comment]
> @hseagle2015 I think this is a “classic ec2” nuance. See [pulumi/pulumi-aws#852](https://github.com/pulumi/pulumi-aws/issues/852) for a discussion about it.
> 
> The fix is to use `vpcSecurityGroupIds` rather than `securityGroups` on the instance.

In Python API reference I don't see an option to use `vpcSecurityGroupIds` nor `securityGroups`, only `security_groups` which I currently use in `pulumi_aws.ec2.Instance` [class](https://www.pulumi.com/docs/reference/pkg/python/pulumi_aws/ec2/#pulumi_aws.ec2.Instance).

Can you give me an example in Python?

**Comment by stack72** [comment]
Hi @hseagle2015 

This should style example should work - notice the use of `vpc_security_group_ids` in the ec2.Instance - this is what @clstokes was talking about

```
import pulumi
import pulumi_aws as aws

size = 't2.micro'

ami = aws.get_ami(most_recent="true",
                  owners=["137112412989"],
                  filters=[{"name":"name","values":["amzn-ami-hvm-*"]}])

group = aws.ec2.SecurityGroup('web-secgrp',
    description='Enable HTTP access',
    ingress=[
        { 'protocol': 'tcp', 'from_port': 80, 'to_port': 80, 'cidr_blocks': ['0.0.0.0/0'] }
    ])

user_data = """
#!/bin/bash
echo "Hello, World!" > index.html
nohup python -m SimpleHTTPServer 80 &
"""

server = aws.ec2.Instance('web-server-www',
    instance_type=size,
    vpc_security_group_ids=[group.id],
    user_data=user_data,
    ami=ami.id)

pulumi.export('public_ip', server.public_ip)
pulumi.export('public_dns', server.public_dns)
```

Paul

**Comment by hseagle2015** [comment]
@stack72 thank you very much. This resolved the problem I was dealing with.
